File: week4/4-1.py
# [21년 재직자 대회] 코딩 테스트 세트 

def test(mid):
    slist = [0] * N
    slist[0] = clist[0]
    
    for i in range(N-1):
        if mid <= slist[i]:
            slist[i+1] = clist[i+1] + dlist[i]
        elif mid <= slist[i] + dlist[i]:
            slist[i+1] = clist[i+1] + (slist[i] + dlist[i] - mid)
        else:
            return False
        
    return True if mid <= slist[N-1] else False

def binary_search(start, end):
    if start == end:
        return start
    
    mid = (start+end+1) // 2
    if(test(mid)):
        return binary_search(mid, end)
    else:
        return binary_search(start, mid-1)

N, T = map(int, input().split())
for i in range(T):
    clist = []
    dlist = []
    
    tmp = list(map(int, input().split()))
    for j in range(len(tmp)):
        if j%2 == 0:
            clist.append(tmp[j])
        else:
            dlist.append(tmp[j])
            
    print(binary_search(0, 2*(10**12)))

File: week4/4-1.cpp
/* [21년 재직자 대회] 코딩 테스트 세트 */

#include <iostream>
#include <vector>
#include <cmath>

using namespace std;

vector<int> clist;
vector<int> dlist;
int N, T;

bool test(int mid){
    vector<int> slist(N,0);
    slist[0] = clist[0];

    for(int i = 0; i < N-1; i++){
        if(mid <= slist[i]){
            slist[i+1] = clist[i+1] + dlist[i];
        }
        else if(mid <= slist[i] + dlist[i]){
            slist[i+1] = clist[i+1] + (slist[i] + dlist[i] - mid);
        }
        else{
            return false;
        }
    }

    return (mid <= slist[N-1]) ? true : false;
}

int binarySearch(int start, int end){
    if(start == end){
        return start;
    }

    int mid = (start+end+1) / 2;
    if(test(mid)){
        return binarySearch(mid, end);
    }   
    else{
        return binarySearch(start, mid-1);
    }
}

int main(){
    cin >> N >> T;

    for(int i = 0; i < T; i++){
        int result = 0;

        for(int j = 0; j < (2*N-1); j++){
            if(j%2 == 0){
                int c;
                cin >> c;
                clist.push_back(c);
            }
            else{
                int d;
                cin >> d;
                dlist.push_back(d);
            }
        }

        cout << binarySearch(0, 2*pow(10,12)) << '\n';
        clist.clear();
        dlist.clear();
    }

    return 0;
}